build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchBuildFinishedPlugin.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.internal;
+
+import com.gradle.scan.plugin.BuildScanExtension;
+
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
+import org.apache.commons.compress.compressors.bzip2.BZip2CompressorOutputStream;
+import org.apache.commons.io.IOUtils;
+import org.elasticsearch.gradle.util.GradleUtils;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.file.FileSystemOperations;
+import org.gradle.api.flow.FlowAction;
+import org.gradle.api.flow.FlowParameters;
+import org.gradle.api.flow.FlowProviders;
+import org.gradle.api.flow.FlowScope;
+import org.gradle.api.internal.file.FileOperations;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.Input;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+public abstract class ElasticsearchBuildFinishedPlugin implements Plugin<Project> {
+
+    @Inject
+    protected abstract FlowScope getFlowScope();
+
+    @Inject
+    protected abstract FlowProviders getFlowProviders();
+
+    @Inject
+    protected abstract FileOperations getFileOperations();
+
+    @Override
+    public void apply(Project target) {
+        String buildNumber = System.getenv("BUILD_NUMBER") != null
+            ? System.getenv("BUILD_NUMBER")
+            : System.getenv("BUILDKITE_BUILD_NUMBER");
+        String performanceTest = System.getenv("BUILD_PERFORMANCE_TEST");
+        if (buildNumber != null && performanceTest == null && GradleUtils.isIncludedBuild(target) == false) {
+            File targetFile = target.file("build/" + buildNumber + ".tar.bz2");
+            File projectDir = target.getProjectDir();
+            File gradleWorkersDir = new File(target.getGradle().getGradleUserHomeDir(), "workers/");
+            BuildScanExtension extension = target.getExtensions().getByType(BuildScanExtension.class);
+            File daemonsLogDir = new File(target.getGradle().getGradleUserHomeDir(), "daemon/" + target.getGradle().getGradleVersion());
+
+            getFlowScope().always(BuildFinishedFlowAction.class, spec -> {
```

Comment:
What on earth is a `FlowAction` and how does this equate to a "build finished"? I can't see anythere here that defines _when_ this action runs.

Reply:
Don't get me started. That is the last half baked overly complicated gradle api that allows code executing on buildFinished. The when is declared implicitly by this line where we map the BuildWorkResult which implicitly defers this code execution till that result is available. There is currently only `getFlowProviders().getBuildWorkResult()` as "flow provider". It's been around for some versions now, obviously incubating but all that is available right now.

```                
spec.getParameters().getFilteredFiles().addAll(getFlowProviders().getBuildWorkResult().map((result) -> {
```

Reply:
So calling `getBuildWorkResult()` implicitly makes this a "build finished" hook. That's not confusing...

Reply:
it is. but yes.